### ISSUE TITLE
Errors when running the fsdp benchmarks for hf_Bert and hf_T5

### ISSUE THREAD

**Issue by 0x6b64** [pytorch/pytorch#93288]
### 🐛 RuntimeError thrown by Dynamo in FSDP benchmark script.

Hi, 

I'm using https://github.com/pytorch/pytorch/blob/master/benchmarks/dynamo/distributed.py to run tests on nightly branch from Jan 28'th (which I've built && installed on my Docker container). 

This is my base DLC: `763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-training:1.13.1-gpu-py39-cu117-ubuntu20.04-sagemaker` (download it using awscli; https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html)

PT-1.13 was purged from the container and PT2 nightly from 28th was installed. 

In particular for `hf_Bert`, `hf_GPT2` and `hf_T5` (the first two with the FSDP wrapper). I'm observing the errors attached below (trace mentions dynamo, so tagging it in the `torch.compile` cohort),

Run command:

```bash
# pytorch commit: 5d6a4f697cac34d15262aad8afab096170d29ce1
torchrun --nnodes 1 --nproc_per_node 8 --master_addr localhost --master_port 9001 distributed.py --dynamo inductor --torchbench_model hf_T5 --fsdp
```

Note commit in the versions below is different because I have some minor change to the build file to include directories to get the build to succeed on my base DLC. 

I've attached the minified graph below, but that doesn't reproduce the issue. 

```bash
TORCHDYNAMO_REPRO_AFTER="dynamo" torchrun --nnodes 1 --nproc_per_node 8 --master_addr localhost --master_port 9001 distributed.py --dynamo inductor --torchbench_model hf_T5 --fsdp
```

Any pointers to get this working OR just any example of working FSDP wrapper would be very helpful!

### Error logs

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 323, in _compile
    out_code = transform_code_object(code, transform)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/bytecode_transformation.py", line 341, in transform_code_object
    transformations(instructions, code_options)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 310, in transform
    tracer.run()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1717, in run
    super().run()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 563, in run
    and self.step()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 526, in step
    getattr(self, inst.opname)(inst)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 332, in wrapper
    return inner_fn(self, inst)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 999, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 460, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/nn_module.py", line 559, in call_function
    return variables.UserFunctionVariable(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/functions.py", line 259, in call_function
    return super(UserFunctionVariable, self).call_function(tx, args, kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/functions.py", line 92, in call_function
    return tx.inline_user_function_return(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 496, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1795, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1851, in inline_call_
    tracer.run()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 563, in run
    and self.step()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 526, in step
    getattr(self, inst.opname)(inst)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1064, in LOAD_ATTR
    result = BuiltinVariable(getattr).call_function(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builtin.py", line 322, in call_function
    result = handler(tx, *args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builtin.py", line 736, in call_getattr
    obj.var_getattr(tx, name).clone(source=source).add_options(options)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/user_defined.py", line 342, in var_getattr
    return VariableBuilder(tx, source)(subobj).add_options(options)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 170, in __call__
    return self._wrap(value).clone(**self.options())
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 236, in _wrap
    return self.wrap_tensor(value)
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 646, in wrap_tensor
    tensor_variable = wrap_fx_proxy(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 761, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 821, in wrap_fx_proxy_cls
    example_value = wrap_to_fake_tensor_and_record(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 972, in wrap_to_fake_tensor_and_record
    fake_e = wrap_fake_exception(
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 783, in wrap_fake_exception
    return fn()
  File "/opt/conda/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 973, in <lambda>
    lambda: tx.fake_mode.from_tensor(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1082, in from_tensor
    return self.fake_tensor_converter(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 291, in __call__
    return self.from_real_tensor(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 248, in from_real_tensor
    out = self.meta_converter(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/meta_utils.py", line 502, in __call__
    r = self.meta_tensor(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/meta_utils.py", line 275, in meta_tensor
    base = self.meta_tensor(
  File "/opt/conda/lib/python3.9/site-packages/torch/_subclasses/meta_utils.py", line 433, in meta_tensor
    r.grad = self.meta_tensor(
RuntimeError: assigned grad has data of a different size

from user code:
   File "/opt/conda/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py", line 542, in <graph break in forward>
    attn_output = self.o(attn_output)
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)

Set torch._dynamo.config.verbose=True for more information
```

### Minified repro

```python
# minifier_launcher.py

import os
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import functools
import torch._dynamo
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch._dynamo.optimizations.backends import BACKENDS
from torch._dynamo.testing import rand_strided

import torch._dynamo.config
import torch._inductor.config
torch._dynamo.config.load_config(b'\x80\x04\x95\xb9\x07\x00\x00\x00\x00\x00\x00}\x94(\x8c\x08__name__\x94\x8c\x14torch._dynamo.config\x94\x8c\x07__doc__\x94N\x8c\x0b__package__\x94\x8c\rtorch._dynamo\x94\x8c\n__loader__\x94\x8c\x1a_frozen_importlib_external\x94\x8c\x10SourceFileLoader\x94\x93\x94)\x81\x94}\x94(\x8c\x04name\x94h\x02\x8c\x04path\x94\x8c>/opt/conda/lib/python3.9/site-packages/torch/_dynamo/config.py\x94ub\x8c\x08__spec__\x94\x8c\x11_frozen_importlib\x94\x8c\nModuleSpec\x94\x93\x94)\x81\x94}\x94(h\x0ch\x02\x8c\x06loader\x94h\n\x8c\x06origin\x94h\x0e\x8c\x0cloader_state\x94N\x8c\x1asubmodule_search_locations\x94N\x8c\r_set_fileattr\x94\x88\x8c\x07_cached\x94\x8cV/opt/conda/lib/python3.9/site-packages/torch/_dynamo/__pycache__/config.cpython-39.pyc\x94\x8c\r_initializing\x94\x89ub\x8c\x08__file__\x94h\x0e\x8c\n__cached__\x94h\x1b\x8c\x07abspath\x94\x8c\tposixpath\x94h\x1f\x93\x94\x8c\x07dirname\x94h h"\x93\x94\x8c\x0eHAS_REFS_PRIMS\x94\x88\x8c\tlog_level\x94K\x1e\x8c\x0boutput_code\x94\x89\x8c\rlog_file_name\x94N\x8c\x07verbose\x94\x89\x8c\x11output_graph_code\x94\x89\x8c\x12verify_correctness\x94\x89\x8c\x12minimum_call_count\x94K\x01\x8c\x15dead_code_elimination\x94\x88\x8c\x10cache_size_limit\x94K@\x8c\x14specialize_int_float\x94\x88\x8c\x0edynamic_shapes\x94\x89\x8c\x10guard_nn_modules\x94\x89\x8c\x0cnormalize_ir\x94\x89\x8c\x1btraceable_tensor_subclasses\x94\x8f\x94\x8c\x0fsuppress_errors\x94\x89\x8c\x15replay_record_enabled\x94\x89\x8c rewrite_assert_with_torch_assert\x94\x88\x8c\x12print_graph_breaks\x94\x89\x8c\x07disable\x94\x89\x8c*allowed_functions_module_string_ignorelist\x94\x8f\x94(\x8c\rtorch.testing\x94\x8c\x0btorch._refs\x94\x8c\rtorch._decomp\x94\x8c\x13torch.distributions\x94\x8c\x0ctorch._prims\x94\x90\x8c\x16capture_scalar_outputs\x94\x89\x8c\x19enforce_cond_guards_match\x94\x88\x8c\x0coptimize_ddp\x94\x88\x8c\x1araise_on_ctx_manager_usage\x94\x88\x8c\x1craise_on_unsafe_aot_autograd\x94\x89\x8c\rdynamo_import\x94\x8c\rtorch._dynamo\x94\x8c\x0finductor_import\x94\x8c\x0ftorch._inductor\x94\x8c\x18error_on_nested_fx_trace\x94\x88\x8c\tallow_rnn\x94\x89\x8c\x08base_dir\x94\x8c&/opt/conda/lib/python3.9/site-packages\x94\x8c\x0edebug_dir_root\x94\x8cN/fsx/pytorch-nightly/pytorch-2.0/pytorch/benchmarks/dynamo/torch_compile_debug\x94\x8c)DO_NOT_USE_legacy_non_fake_example_inputs\x94\x89\x8c\x15_AccessLimitingConfig\x94}\x94(\x8c\n__module__\x94h\x02\x8c\x0b__setattr__\x94h\x02\x8c!_AccessLimitingConfig.__setattr__\x94\x93\x94h\x03Nu\x8c\x15_allowed_config_names\x94\x8f\x94(\x8c\x0eexternal_utils\x94h6\x8c\x12constant_functions\x94h\x06h&\x8c\x0c__builtins__\x94h%hPh5\x8c!skipfiles_inline_module_allowlist\x94h(h\x0fhBh*h/h.h\x1fhD\x8c\x07logging\x94\x8c\x02os\x94\x8c\x05torch\x94h)h$h0h7hIh4h8h@h2h9h-\x8c\x03sys\x94hGh,h+hKhAhEh1hOh\x01hMh\x1dh\'h\x03h"\x8c\x0brepro_level\x94hJ\x8c\nModuleType\x94hCh\x1eh\x04\x8c\x0brepro_after\x94\x90\x8c\x1cget_config_serialization_fns\x94\x8c\x1atorch._dynamo.config_utils\x94hc\x93\x94u.')
torch._inductor.config.load_config(b'\x80\x04\x95\xd3\x08\x00\x00\x00\x00\x00\x00}\x94(\x8c\x08__name__\x94\x8c\x16torch._inductor.config\x94\x8c\x07__doc__\x94N\x8c\x0b__package__\x94\x8c\x0ftorch._inductor\x94\x8c\n__loader__\x94\x8c\x1a_frozen_importlib_external\x94\x8c\x10SourceFileLoader\x94\x93\x94)\x81\x94}\x94(\x8c\x04name\x94h\x02\x8c\x04path\x94\x8c@/opt/conda/lib/python3.9/site-packages/torch/_inductor/config.py\x94ub\x8c\x08__spec__\x94\x8c\x11_frozen_importlib\x94\x8c\nModuleSpec\x94\x93\x94)\x81\x94}\x94(h\x0ch\x02\x8c\x06loader\x94h\n\x8c\x06origin\x94h\x0e\x8c\x0cloader_state\x94N\x8c\x1asubmodule_search_locations\x94N\x8c\r_set_fileattr\x94\x88\x8c\x07_cached\x94\x8cX/opt/conda/lib/python3.9/site-packages/torch/_inductor/__pycache__/config.cpython-39.pyc\x94\x8c\r_initializing\x94\x89ub\x8c\x08__file__\x94h\x0e\x8c\n__cached__\x94h\x1b\x8c\x05debug\x94\x89\x8c\x10disable_progress\x94\x88\x8c\x10verbose_progress\x94\x89\x8c\x0bcpp_wrapper\x94\x89\x8c\x03dce\x94\x89\x8c\x14static_weight_shapes\x94\x88\x8c\x0csize_asserts\x94\x88\x8c\x10pick_loop_orders\x94\x88\x8c\x0finplace_buffers\x94\x88\x8c\x11benchmark_harness\x94\x88\x8c\x0fepilogue_fusion\x94\x89\x8c\x15epilogue_fusion_first\x94\x89\x8c\x0cmax_autotune\x94\x89\x8c\x17realize_reads_threshold\x94K\x04\x8c\x17realize_bytes_threshold\x94M\xd0\x07\x8c\x1brealize_acc_reads_threshold\x94K\x08\x8c\x0ffallback_random\x94\x89\x8c\x12implicit_fallbacks\x94\x88\x8c\rprefuse_nodes\x94\x88\x8c\x0btune_layout\x94\x89\x8c\x11aggressive_fusion\x94\x89\x8c\x0fmax_fusion_size\x94K@\x8c\x1bunroll_reductions_threshold\x94K\x08\x8c\x0ecomment_origin\x94\x89\x8c\tis_fbcode\x94h\x02h7\x93\x94\x8c\x0fcompile_threads\x94K \x8c\x13kernel_name_max_ops\x94K\n\x8c\x0finductor_import\x94\x8c\x0ftorch._inductor\x94\x8c\rshape_padding\x94\x89\x8c\x0epermute_fusion\x94\x89\x8c\x1aprofiler_mark_wrapper_call\x94\x89\x8c\x03cpp\x94}\x94(\x8c\n__module__\x94h\x02\x8c\x07threads\x94J\xff\xff\xff\xff\x8c\x0fdynamic_threads\x94\x89\x8c\x07simdlen\x94N\x8c\x0emin_chunk_size\x94M\x00\x10\x8c\x03cxx\x94N\x8c\x03g++\x94\x86\x94\x8c\x15enable_kernel_profile\x94\x89h\x03Nu\x8c\x06triton\x94}\x94(hBh\x02\x8c\ncudagraphs\x94\x89\x8c\x10debug_sync_graph\x94\x89\x8c\x11debug_sync_kernel\x94\x89\x8c\x0bconvolution\x94\x8c\x04aten\x94\x8c\x0edense_indexing\x94\x89\x8c\tmax_tiles\x94K\x02\x8c\x12autotune_pointwise\x94\x88\x8c tiling_prevents_pointwise_fusion\x94\x88\x8c tiling_prevents_reduction_fusion\x94\x88\x8c\x14ordered_kernel_names\x94\x89\x8c\x18descriptive_kernel_names\x94\x89h\x03Nu\x8c\x05trace\x94}\x94(hBh\x02\x8c\x07enabled\x94\x89\x8c\tdebug_log\x94\x88\x8c\x08info_log\x94\x89\x8c\x08fx_graph\x94\x88\x8c\rir_pre_fusion\x94\x88\x8c\x0eir_post_fusion\x94\x88\x8c\x0boutput_code\x94\x88\x8c\rgraph_diagram\x94\x89\x8c\x0fcompile_profile\x94\x89\x8c\nupload_tar\x94Nh\x03Nu\x8c\x15InductorConfigContext\x94}\x94(hBh\x02\x8c\x0f__annotations__\x94}\x94(\x8c\rstatic_memory\x94\x8c\x08builtins\x94\x8c\x04bool\x94\x93\x94\x8c\x0ematmul_padding\x94hlh+hl\x8c\x12triton_convolution\x94hj\x8c\x03str\x94\x93\x94\x8c\x17rematerialize_threshold\x94hj\x8c\x03int\x94\x93\x94\x8c\x1brematerialize_acc_threshold\x94hsu\x8c\x05_save\x94h\x02\x8c\x1bInductorConfigContext._save\x94\x93\x94\x8c\x06_apply\x94h\x02\x8c\x1cInductorConfigContext._apply\x94\x93\x94\x8c\x08__init__\x94h\x02\x8c\x1eInductorConfigContext.__init__\x94\x93\x94\x8c\t__enter__\x94h\x02\x8c\x1fInductorConfigContext.__enter__\x94\x93\x94\x8c\x08__exit__\x94h\x02\x8c\x1eInductorConfigContext.__exit__\x94\x93\x94h\x03Nu\x8c\x1cget_config_serialization_fns\x94\x8c\x1atorch._dynamo.config_utils\x94h\x84\x93\x94u.')


# REPLACEABLE COMMENT FOR TESTING PURPOSES


args = [((8, 1024), (1024, 1), torch.int64, 'cuda', False), ((32128, 512), (512, 1), torch.float32, 'cuda', True)]
args = [rand_strided(sh, st, dt, dev).requires_grad_(rg) for (sh, st, dt, dev, rg) in args]


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, input_ids : torch.Tensor, self_embed_tokens_weight : torch.Tensor):
        view = input_ids.view(-1, 1024);  input_ids = None
        embedding = torch.nn.functional.embedding(view, self_embed_tokens_weight, None, None, 2.0, False, False);  view = self_embed_tokens_weight = None
        ones = torch.ones(8, 1024)
        to = ones.to(device(type='cuda', index=0));  ones = None
        return (to, embedding)


mod = Repro()

# Setup debug minifier compiler
torch._dynamo.debug_utils.MINIFIER_SPAWNED = True
compiler_fn = BACKENDS["dynamo_minifier_backend"]

dynamo_minifier_backend = functools.partial(
    compiler_fn,
    compiler_name="inductor",
)
opt_mod = torch._dynamo.optimize(dynamo_minifier_backend)(mod)

with torch.cuda.amp.autocast(enabled=False):
    opt_mod(*args)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0a0+git5876d91
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.24.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:58:50)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-4.14.296-222.539.amzn2.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: <redacted>

Nvidia driver version: 470.103.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.0
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.1.4
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] torch==2.0.0a0+git5876d91
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.0a0+4699ef2
[pip3] torchdata==0.6.0a0+a1612ee
[pip3] torchmetrics==0.11.0
[pip3] torchrec-nightly==2023.1.29
[pip3] torchtext==0.15.0a0+f653dac
[pip3] torchvision==0.15.0a0+c35e8d5
[pip3] vector-quantize-pytorch==0.10.15
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.1.4                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] mkl-include               2023.0.0         h84fe81f_25396    conda-forge
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] torch                     2.0.0a0+git5876d91          pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.0a0+4699ef2          pypi_0    pypi
[conda] torchdata                 0.6.0a0+a1612ee          pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchrec-nightly          2023.1.29                pypi_0    pypi
[conda] torchtext                 0.15.0a0+f653dac          pypi_0    pypi
[conda] torchvision               0.15.0a0+c35e8d5          pypi_0    pypi
[conda] vector-quantize-pytorch   0.10.15                  pypi_0    pypi
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @kiukchung @lucasllc @d4l3k @soumith @ngimel @Xia-Weiwen @desertfire

**Comment by wconstab** [comment]
Just as a datapoint, I've run these commands locally on a build from 1/30 (master @ 2fc73622f8e832179e275239506d749dd30767b3)

It seems like DDP works fine and FSDP runs ok with 1 proc but crashes with multiple proc.

`python benchmarks/dynamo/distributed.py --torchbench_model hf_T5 --ddp --dynamo  inductor`
gave some warnings about Triton random, then `mean latency 0.15712912613525987 across 10 runs`

`torchrun --nproc_per_node 2 benchmarks/dynamo/distributed.py --torchbench_model hf_T5 --ddp --dynamo inductor`
same warnings, and then 
```
mean latency 0.15981919057667254 across 10 runs
mean latency 0.15921848462894558 across 10 runs
```

`python benchmarks/dynamo/distributed.py --torchbench_model hf_T5 --fsdp --dynamo  inductor`
gave some additional warnings like:

```
[2023-01-30 18:28:59,841] torch._dynamo.symbolic_convert: [WARNING] /scratch/whc/work/py38/lib/python3.8/site-packages/transformers/models/t5/modeling_t5.py <function T5Attention._rela
tive_position_bucket at 0x7f4204878040> [UnspecializedNNModuleVariable(T5Attention), TensorVariable()] {'bidirectional': ConstantVariable(bool), 'num_buckets': ConstantVariable(int), '
max_distance': ConstantVariable(int)} multiple values for argument 'bidirectional'
```
and then `mean latency 0.6697056834585965 across 10 runs`

`torchrun --nproc_per_node 2 benchmarks/dynamo/distributed.py --torchbench_model hf_T5 --fsdp --dynamo  inductor`
gives me a backtrace with
`RuntimeError: assigned grad has data of a different size`

**Comment by wconstab** [comment]
We actually test a very similar config in CI which is running and passing on 4gpu.
https://github.com/pytorch/pytorch/blob/master/test/distributed/test_dynamo_distributed.py#L335

I am guessing the issue has to do with one of the settings that the test case specifies  - either disabling inductor's cudagraphs, or, using fallback random

```
    @patch.object(torch._inductor.config.triton, "cudagraphs", False)
    @patch.object(torch._inductor.config, "fallback_random", True)
```

nope, i'm wrong. cudagraphs is already being disabled in the distributed.py benchmark and the fallback_random thing doesn't matter.  

It seems to be a hf_Bert vs hf_T5 issue.  Bert runs fine for me in both unit test and benchmark script.

If you don't care which model you're using and just want to run FSDP as the issue suggests, can you try hf_Bert?  We can look into the hf_T5 issue.

**Comment by 0x6b64** [comment]
Thanks for the response @wconstab . hf_Bert and hf_T5 only show errors with the fsdp wrapper. However, I've run in to errors with DDP with other models as well. I'll be running them again today and add the full list of models for where there is an error in the distributed setting. (ddp/fsdp - though fsdp only seems to have two models with wrapper blocks).

**Comment by wconstab** [comment]
interesting. could you file a separate ticket for the ddp errors and make this one focus on the fsdp ones?

**Comment by 0x6b64** [comment]
Absolutely! I'll refocus this one to just fsdp.

**Comment by agrizzli** [comment]
I experience similar error `RuntimeError: assigned grad has data of a different size` when running a training with FSDP on 8 A100 GPUs.

I tried to train HuggingFace models using Accelerate such as `bert-base-uncased`, `roberta-base`, `xlm-roberta-base` and `facebook/xlm-roberta-xxl`. All with the same result.

I use `use_orig_params = True` to initialize `FSDP` via Accelerate's `FullyShardedDataParallelPlugin` class. Without this option Dynamo fails in general and ask to provided. However, with this option it fails with the above message in the `backward()` pass within the training loop. The relevant part of traceback is:

```
│ /opt/conda/lib/python3.8/site-packages/accelerate/accelerator.py:1683 in     │
│ backward                                                                     │
│                                                                              │
│   1680 │   │   elif self.scaler is not None:                                 │
│   1681 │   │   │   self.scaler.scale(loss).backward(**kwargs)                │
│   1682 │   │   else:                                                         │
│ ❱ 1683 │   │   │   loss.backward(**kwargs)                                   │
│   1684 │                                                                     │
│   1685 │   def unscale_gradients(self, optimizer=None):                      │
│   1686 │   │   """                                                           │
│                                                                              │
│ /opt/conda/lib/python3.8/site-packages/torch/_tensor.py:488 in backward      │
│                                                                              │
│    485 │   │   │   │   create_graph=create_graph,                            │
│    486 │   │   │   │   inputs=inputs,                                        │
│    487 │   │   │   )                                                         │
│ ❱  488 │   │   torch.autograd.backward(                                      │
│    489 │   │   │   self, gradient, retain_graph, create_graph, inputs=inputs │
│    490 │   │   )                                                             │
│    491                                                                       │
│                                                                              │
│ /opt/conda/lib/python3.8/site-packages/torch/autograd/__init__.py:204 in     │
│ backward                                                                     │
│                                                                              │
│   201 │   # The reason we repeat the same comment below is that              │
│   202 │   # some Python versions print out the first line of a multi-line fu │
│   203 │   # calls in the traceback and some print out the last line          │
│ ❱ 204 │   Variable._execution_engine.run_backward(  # Calls into the C++ eng │
│   205 │   │   tensors, grad_tensors_, retain_graph, create_graph, inputs,    │
│   206 │   │   allow_unreachable=True, accumulate_grad=True)  # Calls into th │
│   207                                                                        │
│                                                                              │
│ /opt/conda/lib/python3.8/site-packages/torch/utils/_contextlib.py:115 in     │
│ decorate_context                                                             │
│                                                                              │
│   112 │   @functools.wraps(func)                                             │
│   113 │   def decorate_context(*args, **kwargs):                             │
│   114 │   │   with ctx_factory():                                            │
│ ❱ 115 │   │   │   return func(*args, **kwargs)                               │
│   116 │                                                                      │
│   117 │   return decorate_context                                            │
│   118                                                                        │
│                                                                              │
│ /opt/conda/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils │
│ .py:700 in _post_backward_hook                                               │
│                                                                              │
│    697 │   │                                                                 │
│    698 │   │   if not state._sync_gradients:                                 │
│    699 │   │   │   if handle._use_orig_params:                               │
│ ❱  700 │   │   │   │   handle._use_unsharded_grad_views()                    │
│    701 │   │   │   return                                                    │
│    702 │   │                                                                 │
│    703 │   │   # Wait for all ops in the current stream (e.g. gradient       │
│                                                                              │
│ /opt/conda/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py: │
│ 1686 in _use_unsharded_grad_views                                            │
│                                                                              │
│   1683 │   │   │   │   │   param.grad = torch.empty_like(param)              │
│   1684 │   │   │   │   param.grad.data = view                                │
│   1685 │   │   │   else:                                                     │
│ ❱ 1686 │   │   │   │   param.grad = view                                     │
│   1687 │   │   for i, (                                                      │
│   1688 │   │   │   param_name,                                               │
│   1689 │   │   │   module,                                                   │
╰──────────────────────────────────────────────────────────────────────────────╯
RuntimeError: assigned grad has data of a different type
```

**Comment by wconstab** [comment]
Hey @rohan-varma @awgu could you take a quick look and see if this error is related to mixed-precision support landed by @rohan-varma last year? https://github.com/pytorch/pytorch/pull/85223/files?

It sounds pretty clear from the error that FSDP and the module aren't in agreement about the dtype.

Can @agrizzli or @0x6b64 confirm the expectation on the dtype settings for the grads?

**Comment by awgu** [comment]
The original error reported @0x6b64 comes from a "bug" in the composition of FSDP and Dynamo.

Eager FSDP stretches what is allowed in PyTorch because for the first time, we have parameters and gradients switching between unsharded and sharded states as well as full and low precisions. As such, inside our implementation, we have to use some hacks to bypass existing PyTorch checks while we temporarily violate the invariants guaranteed by those checks.

For example, (1) we may have to unshard a parameter while its gradient is sharded, causing a mismatch in their shapes, which is normally prohibited, or (2) we may have to hold the parameter in full precision while its gradient is in low precision, causing a mismatch in their dtypes. For (1), FSDP ensures that we reshard the parameter before the optimizer step to restore the invariant that the parameter and its gradient have matching shapes. For (2), we actually permit computation ops involving a full precision parameter and low precision gradient since the op will perform type promotion automatically.

For this issue, there is some challenge about what Dynamo should do when it tries tracing an FSDP model that is currently in one of the temporary invalid states (either shape or dtype mismatch between parameters and gradients). I would need to dig more into exactly what conditions need to be met for Dynamo to be tracing and see the mismatched parameter/gradient shapes. If you can `zero_grad(set_to_none=True)`, then that might unblock for now.

For @agrizzli's error, I am not as sure because that is an error directly in FSDP's code, not in Dynamo. Does the error reproduce if you disable `torch.compile()`?

**Comment by agrizzli** [comment]
@wconstab I have checked the `shape` and `dtype` of tensors at line `1686` in `torch/distributed/fsdp/flat_param.py` using 4 GPUs and `param.grad` and `view` are of same type and shape, which is not surprising, since the `else` clause corresponds to `if param.shape != view.shape or param.dtype != view.dtype`. The output ist the following:

```
rank 0 - param.grad: <class 'torch.Tensor'> torch.Size([768]) torch.float32, view: <class 'torch.Tensor'> torch.Size([768]) torch.float32
rank 1 - param.grad: <class 'torch.Tensor'> torch.Size([768]) torch.float32, view: <class 'torch.Tensor'> torch.Size([768]) torch.float32
rank 2 - param.grad: <class 'torch.Tensor'> torch.Size([3072]) torch.float32, view: <class 'torch.Tensor'> torch.Size([3072]) torch.float32
rank 3 - param.grad: <class 'torch.Tensor'> torch.Size([768]) torch.float32, view: <class 'torch.Tensor'> torch.Size([768]) torch.float32
```

There is only a difference between processes in the tensor shape: 768 vs. 3072. I train in full precision using `tf32`. Also with `fp32` the result is the same.

Data inside of `params.grad` is all zero at this point in time for all ranks. `view` is filled with some data.

@awgu Indeed, the error happens also without `torch.compile()` as soon as I use `(FSDP(use_orig_params=True))`. 

I also tried using `zero_grad(set_to_none=True)`, but this does not change the outcome. Interestingly, the `params.grad` was still all zeros, where I was supposing to see `None`. But maybe just wrong understanding from my side here.

**Comment by agrizzli** [comment]
@awgu Done #98494

**Comment by chauhang** [comment]
@agrizzli Can you please test with the latest PT nightlies and share an update for this?

**Comment by agrizzli** [comment]
> @agrizzli Can you please test with the latest PT nightlies and share an update for this?

The sub-issue for a different bug I reported here as #98494  was successfully closed and I don't get the same error anymore.

However, I face problems re-creating the environment for the original (different) issue reported by @0x6b64. Maybe @0x6b64 can be helpful here.

**Comment by anijain2305** [comment]
Closing this issues because of staleness and unable to repro.
